mmgen/models/architectures/biggan/biggan_snmodule.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# yapf:disable
+'''
+    Ref: Functions in this file is borrowed from https://github.com/ajbrock/BigGAN-PyTorch/blob/master/layers.py # noqa
+'''
+# yapf:enable
+
+
+def proj(x, y):
+    """Calculate Projection of x onto y.
+
+    Args:
+        x (torch.Tensor): Projection vector x.
+        y (torch.Tensor): Direction vector y.
+
+    Returns:
+        torch.Tensor: Projection of x onto y.
+    """
+    return torch.mm(y, x.t()) * y / torch.mm(y, y.t())
+
+
+def gram_schmidt(x, ys):
+    """Orthogonalize x wrt list of vectors ys.
+
+    Args:
+        x (torch.Tensor): Vector to be added into the
+            orthogonal vectors.
+        ys (list[torch.Tensor]): A set of orthogonal
+            vectors.
+
+    Returns:
+        torch.Tensor: Result of Gram–Schmidt
+            orthogonalization.
+    """
+    for y in ys:
+        x = x - proj(x, y)
+    return x
+
+
+def power_iteration(W, u_, update=True, eps=1e-12):
+    """Power iteration method for calculating spectral norm.
+
+    Args:
+        W (torch.Tensor): Module weight.
+        u_ (list[torch.Tensor]): list of left singular vector.
```

Comment:
You may revise the copied codes from another repo. This name is not so good.

Reply:
`W` --> `weight`, `u_` --> `u_list`.